On the way to beautiful errors

rcoh opened the pull request:
![Selection_011](https://user-images.githubusercontent.com/492903/84194538-278d0f80-aa6b-11ea-83dc-2242bba5a206.png)
This diff introduces `ErrorWithContext` which is an error message that can print nicely along with attached context from code. Eventually, this will output JSON errors in JSON output mode, but that isn't fully done right now. I use it in a couple of places -- a full conversion of all applicable errors into `ErrorWithContext` is coming in a later diff.

Finally getting some of the payoff of tracking the spans. Eventually, I want to refactor the error handling in `validate_configs` to return the errors instead of just logging them. For now though, I decided to just preserve the existing behavior, there are enough moving parts in this diff already.

We track span contents in a global singleton so we can look up the code a span came from posthoc.

I'm torn if the tests should include the escape sequences or not -- it's good to test that the colors don't suddenly change, but it's definitely harder to verify alignment, eg.

rcoh committed:
On the way to beautiful errors

brendongo commented on semgrep/tests/e2e/snapshots/test_rule_parser/test_rule_parser__failure__error_messages/bad2/error.txt:
```diff
@@ -1,2 +1,12 @@
-rules/syntax/bad2.yaml has an invalid top-level rule key {'pattern-inside'} at rule id eqeq-is-bad, can only have: ['equivalences', 'fix', 'id', 'languages', 'message', 'metadata', 'paths', 'pattern', 'pattern-either', 'pattern-regex', 'patterns', 'severity']
+[31merror[39m: extra top-level key
+  --> rules/syntax/bad2.yaml:3
+[94m1 | [39mrules:
```
Will the color codes exist if a user is piping output to a file?

rcoh replied:
yeah, although this isn't a change how things currently behave se -- In the original version of this diff I stripped all the colors out if it the input was a tty. I can re-add it if you want.

brendongo replied:
I think we should be consistent. I remember a previous PR stripping out color codes when outputing to a file or was that just an early version of this PR?

rcoh replied:
I've readded output stripping & test with and without color.

brendongo commented on semgrep/tests/e2e/rules/syntax/missing-toplevel.yaml:
```diff
@@ -0,0 +1,14 @@
+rule:
```
Can you add a comment/modify the message to mention what top level key is missing

brendongo commented on semgrep/tests/e2e/rules/syntax/missing-field.yaml:
```diff
@@ -0,0 +1,13 @@
+        flask.response.set_cookie(..., httponly=True, secure=True,...)
+    - pattern: |
+        flask.response.set_cookie(....)
+  message: |
```
Can you add a comment/modify message field to mention what field is missing

brendongo commented on semgrep/semgrep/semgrep_main.py:
```diff
@@ -43,14 +45,27 @@ def validate_single_rule(config_id: str, rule_yaml: YamlTree) -> Optional[Rule]:
     rule_keys = set(rule.keys())
     if not rule_keys.issuperset(YAML_MUST_HAVE_KEYS):
         missing_keys = YAML_MUST_HAVE_KEYS - rule_keys
+        # TODO: return the error messages so we can emit nice JSON errors
```
Can you create Issues for TODOs instead of having an inline todo

brendongo commented on semgrep/semgrep/error.py:
```diff
@@ -32,3 +44,108 @@ class InvalidPatternNameError(SemgrepError):
     pass
+
+
+class ErrorWithContext(SemgrepError):
```
Better Named SemgrepRuleFormatError or something more specific. Current name implies it can be any error with context but it only applies to rule format errors.

rcoh replied:
I'm also going to use the same error to display errors from semgrep-core in the code

rcoh replied:
maybe `ErrorWithSpan`?

brendongo replied:
what kind of errors from semgrep-core?

rcoh replied:
invalid patterns & parse errors on ASTs

brendongo commented on semgrep/semgrep/error.py:
```diff
@@ -32,3 +44,108 @@ class InvalidPatternNameError(SemgrepError):
+        self,
+        short_msg: str,
+        long_msg: Optional[str],
+        level: str,
```
Top level comment on what each arg means would be helpful

rcoh committed:
Assorted Improvements

- Betters docs for ErrorWithSpan
- Suppress color by default when the output is not a tty but enable it with `--force-color`
- Add snapshot testing with and without color

brendongo commented:
Can we have more function level comments in general. It's not immediately clear what `emit` or `_code_segment` does.

rcoh committed:
More cleanups

- Add inline comments
- Delete unused emit method
- Change `emit_str` to __repr__ which is hopefully clearer

rcoh committed:
Add lots more comments, change Positions to being 1-indexed
